server/manageuser.php

```diff
@@ -0,0 +1,153 @@
+<?php
+
+require_once 'config.php';
+require_once "user.class.php";
+
+$USER = new User();
+if (!$USER->hasPermission('edit_user_permission')) {
+	header("Location: /unauthorized");
+	die();
+}
+
+if(isset($_REQUEST['callback'])){
+	switch ($_REQUEST['callback']) {
+		case 'getUserTable':
+			getUserTable($_REQUEST);
+			break;
+		case 'updateUserPermissions':
+			updateUserPermissions($_REQUEST);
+			break;
+		case 'addUser':
+			addUser($_REQUEST);
+			break;
+		default:
+			die('Unregistered command. This instance has been reported.');
+			break;
+	}
+}
+
+function getUserTable($data){
+	GLOBAL $DB_CONN;
+
+	$response = array();
+	$response['success'] = true;
+
+	$stmt = $DB_CONN->prepare("SELECT userId, firstName, lastName, email, preparesTaxes, archived 
+		FROM vita.user
+		ORDER BY firstName");
+
+	$stmt->execute(array());
+
+	$thead = "<thead><tr><th>Name</th><th>Email</th><th>Permissions</th><th>Count</th></tr>";
+	$tbody = "<tbody>";
+	while($row = $stmt->fetch(PDO::FETCH_ASSOC)){
+		$optionList = getUserPermissionOptionList($row['userId']);
+
+		$tbody.= "<tr data-user-id='".$row['userId']."'>";
+		$tbody.= "<td>".$row['firstName']." ".$row['lastName']."</td>";
+		$tbody.= "<td>".$row['email']."</td>";
+		$tbody.= "<td><select class='userPermissionList selectpicker' multiple=true>".implode('', $optionList['options'])."</select></td>";
+		$tbody.= "<td>".$optionList['hasPermissionCount']."</td>";
+		$tbody.= "</tr>";
+	}
+	$tbody.= "</tbody>";
+	$thead.= "</tr></thead>";
+
+	$response['table'] = $thead.$tbody;
+
+	print json_encode($response);
+}
+
+// Returns string of html <option> elments with all permissions
+function getUserPermissionOptionList($userId){
+	GLOBAL $DB_CONN;
+
+	$stmt = $DB_CONN->prepare("SELECT 
+		permissionId,
+		name,
+        description,
+        lookupName,
+        (SELECT COUNT(*) FROM userpermission WHERE userId = ? AND permissionId = p.permissionId) as hasPermission
+	FROM permission p");
+
+	$stmt->execute(array($userId));
+
+	$options = array();
+
+	$count = 0;
+	while($row = $stmt->fetch(PDO::FETCH_ASSOC)){
+		if($row['hasPermission']){
+			$selected = 'selected=true';
+			$count++;
+		}else{
+			$selected = '';
+		}
+		array_push($options, "<option value=".$row['permissionId']." $selected>".$row['name']."</option>");
+	}
+
+	return array(
+		'options' => $options,
+		'hasPermissionCount' => $count
+	);
+}
+
+function updateUserPermissions($data){
+	GLOBAL $DB_CONN;
+	GLOBAL $USER;
+
+	$response = array();
+	$response['success'] = true;
+
+	$DB_CONN->beginTransaction();
+
+	$stmt = $DB_CONN->prepare("DELETE FROM UserPermission 
```

Comment:
I'm not sure how I feel about how this method works. Correct me if I'm wrong, but it appears that you delete all the permissions they have in the database, and then add the ones that come in through $data. Is there any chance that we could change how we pass the data so we can see which ones need to be deleted and which ones need to be added? This way, we're not deleting all the permissions they have and then re-adding them. This way the audit trail of "createdBy" will be correct, and not just the last permission who edited the permissions.

Furthermore, I think this is a source of a bug. If a person removes/adds a permission to themselves, then it returns a 404 with an unauthorized. Then, the user can't navigate back to the /manage/user page.

Reply:
I fixed the first part but it's still the source of a bug. Should I prevent users from removing their own permission that gives them permission to edit permissions?